File: app/src/main/java/com/jojokorok/nourishrx/ui/AppShellFlow.java
package com.jojokorok.nourishrx.ui;

import android.app.Activity;
import android.graphics.Color;
import android.graphics.Typeface;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.jojokorok.nourishrx.data.Medication;
import com.jojokorok.nourishrx.data.MedicationStore;
import com.jojokorok.nourishrx.data.Profile;
import com.jojokorok.nourishrx.data.WeightEntry;

import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Locale;

public class AppShellFlow {
    public interface Callbacks {
        long currentProfileId();

        String currentMode();

        String currentTab();

        Profile selectedProfile();

        View profileAvatar(Profile profile, int sizeDp, int fallbackColor, int textSp);

        int avatarWidthDp(Profile profile, int heightDp);

        int todayDoseCount();

        String alertsLabel();

        int alertColor();

        void showQuickAdd(boolean nutritionMode);

        void showProfiles();

        void selectAbout();

        void selectMode(String mode);

        void selectTab(String tab);

        void handleAlertsTap();
    }

    private static final String MODE_MEDICATION = "medication";
    private static final String MODE_NUTRITION = "nutrition";
    private static final String TAB_ABOUT = "about";

    private final Activity activity;
    private final MedicationStore store;
    private final NourishUi ui;
    private final ZoneId zoneId;
    private final Callbacks callbacks;
    private final DateTimeFormatter dateFormatter =
            DateTimeFormatter.ofPattern("EEEE, MMM d", Locale.getDefault());

    public AppShellFlow(
            Activity activity,
            MedicationStore store,
            NourishUi ui,
            ZoneId zoneId,
            Callbacks callbacks
    ) {
        this.activity = activity;
        this.store = store;
        this.ui = ui;
        this.zoneId = zoneId;
        this.callbacks = callbacks;
    }

    public LinearLayout render() {
        LinearLayout root = new LinearLayout(activity);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setBackgroundColor(NourishColors.SURFACE);
        root.setPadding(ui.dp(16), ui.dp(12), ui.dp(16), 0);

        root.addView(headerPanel());
        root.addView(tabRow());

        ScrollView scrollView = new ScrollView(activity);
        scrollView.setFillViewport(true);
        LinearLayout content = new LinearLayout(activity);
        content.setOrientation(LinearLayout.VERTICAL);
        content.setPadding(0, ui.dp(14), 0, ui.dp(24));
        scrollView.addView(content, new ScrollView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        ));
        root.addView(scrollView, new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                0,
                1
        ));

        activity.setContentView(root);
        return content;
    }

    private View headerPanel() {
        LinearLayout panel = new LinearLayout(activity);
        panel.setOrientation(LinearLayout.VERTICAL);
        panel.setPadding(ui.dp(16), ui.dp(16), ui.dp(16), ui.dp(14));
        panel.setBackground(ui.roundedGradient(
                new int[]{
                        Color.rgb(222, 244, 231),
                        Color.rgb(255, 237, 215),
                        Color.rgb(236, 242, 255)
                },
                ui.dp(26)
        ));
        panel.setElevation(ui.dp(2));

        LinearLayout top = new LinearLayout(activity);
        top.setOrientation(LinearLayout.HORIZONTAL);
        top.setGravity(Gravity.CENTER_VERTICAL);

        Profile profile = callbacks.selectedProfile();
        View mark = callbacks.profileAvatar(profile, 54, NourishColors.GREEN, 18);
        LinearLayout.LayoutParams markParams = new LinearLayout.LayoutParams(
                ui.dp(callbacks.avatarWidthDp(profile, 54)),
                ui.dp(54)
        );
        markParams.rightMargin = ui.dp(12);
        top.addView(mark, markParams);

        LinearLayout titleGroup = new LinearLayout(activity);
        titleGroup.setOrientation(LinearLayout.VERTICAL);
        TextView headline = ui.displayText(profile.name, 28, NourishColors.INK);
        headline.setSingleLine(true);
        headline.setEllipsize(TextUtils.TruncateAt.END);
        titleGroup.addView(headline);
        titleGroup.addView(ui.text(
                "Today - " + LocalDate.now().format(dateFormatter),
                13,
                NourishColors.MUTED,
                Typeface.BOLD
        ));
        top.addView(titleGroup, new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1));

        boolean nutritionMode = MODE_NUTRITION.equals(callbacks.currentMode());
        Button add = ui.button(nutritionMode ? "+ Log" : "+ Med", Color.WHITE, NourishColors.GREEN);
        add.setOnClickListener(view -> callbacks.showQuickAdd(nutritionMode));
        top.addView(add, compactButtonParams());
        panel.addView(top);

        LinearLayout utilityActions = new LinearLayout(activity);
        utilityActions.setOrientation(LinearLayout.HORIZONTAL);

        Button profileButton = ui.button("Manage profiles", NourishColors.BLUE, Color.WHITE);
        profileButton.setOnClickListener(view -> callbacks.showProfiles());
        utilityActions.addView(profileButton, weightedActionParams());

        boolean showingAbout = TAB_ABOUT.equals(callbacks.currentTab());
        Button aboutButton = ui.button(
                "About",
                showingAbout ? Color.WHITE : NourishColors.BLUE,
                showingAbout ? NourishColors.BLUE : Color.WHITE
        );
        aboutButton.setOnClickListener(view -> callbacks.selectAbout());
        utilityActions.addView(aboutButton, new LinearLayout.LayoutParams(0, ui.dp(44), 1));

        LinearLayout.LayoutParams utilityParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ui.dp(44)
        );
        utilityParams.topMargin = ui.dp(12);
        panel.addView(utilityActions, utilityParams);

        LinearLayout.LayoutParams modeParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ui.dp(46)
        );
        modeParams.topMargin = ui.dp(10);
        panel.addView(modeSwitchRow(), modeParams);

        panel.addView(headerStats(nutritionMode));

        Button alerts = ui.button(callbacks.alertsLabel(), callbacks.alertColor(), Color.WHITE);
        alerts.setOnClickListener(view -> callbacks.handleAlertsTap());
        LinearLayout.LayoutParams alertParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ui.dp(44)
        );
        alertParams.topMargin = ui.dp(12);
        panel.addView(alerts, alertParams);

        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        params.bottomMargin = ui.dp(12);
        panel.setLayoutParams(params);
        return panel;
    }

    private LinearLayout headerStats(boolean nutritionMode) {
        LinearLayout stats = new LinearLayout(activity);
        stats.setOrientation(LinearLayout.HORIZONTAL);
        stats.setPadding(0, ui.dp(14), 0, 0);
        if (nutritionMode) {
            LocalDate today = LocalDate.now(zoneId);
            long start = today.atStartOfDay(zoneId).toInstant().toEpochMilli();
            long end = today.plusDays(1).atStartOfDay(zoneId).toInstant().toEpochMilli();
            int foodLogCount = store.getMealFoodLogs(callbacks.currentProfileId(), start, end).size();
            int waterOunces = store.getWaterOunces(callbacks.currentProfileId(), start, end);
            List<WeightEntry> weights = store.getWeightEntries(callbacks.currentProfileId(), 1);
            stats.addView(ui.summaryPill(
                    plural(foodLogCount, "food log", "food logs"),
                    NourishColors.GREEN,
                    NourishColors.GREEN_SOFT
            ));
            stats.addView(ui.summaryPill(
                    waterOunces + " oz water",
                    NourishColors.BLUE,
                    NourishColors.BLUE_SOFT
            ));
            stats.addView(ui.summaryPill(
                    weights.isEmpty() ? "no weight" : formatValue(weights.get(0).pounds) + " lb",
                    NourishColors.GOLD,
                    NourishColors.GOLD_SOFT
            ));
        } else {
            List<Medication> medications = store.getAllMedications(callbacks.currentProfileId());
            long lowCount = medications.stream().filter(Medication::isLowStock).count();
            stats.addView(ui.summaryPill(
                    plural(callbacks.todayDoseCount(), "dose", "doses"),
                    NourishColors.GREEN,
                    NourishColors.GREEN_SOFT
            ));
            stats.addView(ui.summaryPill(
                    plural(medications.size(), "med", "meds"),
                    NourishColors.BLUE,
                    NourishColors.BLUE_SOFT
            ));
            stats.addView(ui.summaryPill(
                    plural(lowCount, "refill", "refills"),
                    NourishColors.GOLD,
                    NourishColors.GOLD_SOFT
            ));
        }
        return stats;
    }

    private LinearLayout modeSwitchRow() {
        LinearLayout modes = new LinearLayout(activity);
        modes.setOrientation(LinearLayout.HORIZONTAL);
        modes.setPadding(ui.dp(4), ui.dp(4), ui.dp(4), ui.dp(4));
        modes.setBackground(ui.rounded(NourishColors.CARD, NourishColors.BORDER, ui.dp(20)));
        modes.addView(modeButton("Medication", MODE_MEDICATION));
        modes.addView(modeButton("Nutrition", MODE_NUTRITION));
        return modes;
    }

    private Button modeButton(String label, String mode) {
        boolean selected = callbacks.currentMode().equals(mode);
        Button button = ui.button(
                label,
                selected ? Color.WHITE : NourishColors.MUTED,
                selected ? NourishColors.BLUE : Color.TRANSPARENT
        );
        button.setOnClickListener(view -> callbacks.selectMode(mode));
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(0, ui.dp(38), 1);
        params.leftMargin = ui.dp(2);
        params.rightMargin = ui.dp(2);
        button.setLayoutParams(params);
        return button;
    }

    private LinearLayout tabRow() {
        LinearLayout tabs = new LinearLayout(activity);
        tabs.setOrientation(LinearLayout.HORIZONTAL);
        tabs.setPadding(ui.dp(4), ui.dp(4), ui.dp(4), ui.dp(4));
        tabs.setBackground(ui.rounded(NourishColors.TAB_TRACK, Color.TRANSPARENT, ui.dp(20)));
        if (MODE_NUTRITION.equals(callbacks.currentMode())) {
            tabs.addView(tabButton("Today", "nutrition_today"));
            tabs.addView(tabButton("Meals", "nutrition_meals"));
            tabs.addView(tabButton("Saved", "nutrition_saved"));
            tabs.addView(tabButton("Foods", "nutrition_foods"));
            tabs.addView(tabButton("Body", "nutrition_body"));
        } else {
            tabs.addView(tabButton("Today", "today"));
            tabs.addView(tabButton("Meds", "meds"));
            tabs.addView(tabButton("Stock", "stock"));
        }
        return tabs;
    }

    private Button tabButton(String label, String tab) {
        boolean selected = callbacks.currentTab().equals(tab);
        Button button = ui.button(
                label,
                selected ? Color.WHITE : NourishColors.MUTED,
                selected ? NourishColors.GREEN : Color.TRANSPARENT
        );
        button.setTextSize(12);
        button.setSingleLine(true);
        button.setMaxLines(1);
        button.setEllipsize(TextUtils.TruncateAt.END);
        button.setMinWidth(0);
        button.setMinimumWidth(0);
        button.setPadding(ui.dp(4), 0, ui.dp(4), 0);
        button.setOnClickListener(view -> callbacks.selectTab(tab));
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(0, ui.dp(42), 1);
        params.leftMargin = ui.dp(2);
        params.rightMargin = ui.dp(2);
        button.setLayoutParams(params);
        return button;
    }

    private LinearLayout.LayoutParams compactButtonParams() {
        return new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ui.dp(44)
        );
    }

    private LinearLayout.LayoutParams weightedActionParams() {
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(0, ui.dp(44), 1);
        params.rightMargin = ui.dp(8);
        return params;
    }

    private String plural(long count, String singular, String plural) {
        return count + " " + (count == 1 ? singular : plural);
    }

    private String formatValue(float value) {
        if (Math.abs(value - Math.round(value)) < 0.05f) {
            return String.valueOf(Math.round(value));
        }
        return String.format(Locale.getDefault(), "%.1f", value);
    }
}

File: app/src/main/java/com/jojokorok/nourishrx/ui/AppShellFlow.kt
package com.jojokorok.nourishrx.ui

import android.app.Activity
import android.graphics.Color
import android.graphics.Typeface
import android.os.Build
import android.text.TextUtils
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.WindowInsets
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import com.jojokorok.nourishrx.R
import com.jojokorok.nourishrx.data.MedicationStore
import com.jojokorok.nourishrx.data.Profile
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToInt

class AppShellFlow(
    private val activity: Activity,
    private val store: MedicationStore,
    private val ui: NourishUi,
    private val zoneId: ZoneId,
    private val callbacks: Callbacks
) {
    interface Callbacks {
        fun currentProfileId(): Long
        fun currentMode(): String
        fun currentTab(): String
        fun selectedProfile(): Profile
        fun profileAvatar(profile: Profile, sizeDp: Int, fallbackColor: Int, textSp: Int): View
        fun avatarWidthDp(profile: Profile, heightDp: Int): Int
        fun todayDoseCount(): Int
        fun alertsLabel(): String
        fun alertColor(): Int
        fun showQuickAdd(nutritionMode: Boolean)
        fun showProfiles()
        fun selectAbout()
        fun selectMode(mode: String)
        fun selectTab(tab: String)
        fun handleAlertsTap()
    }

    private data class HeaderStat(
        val value: String,
        val label: String,
        val color: Int
    )

    private val dateFormatter = DateTimeFormatter.ofPattern("EEEE, MMMM d", Locale.getDefault())

    fun render(): LinearLayout {
        val nutritionMode = callbacks.currentMode() == MODE_NUTRITION
        val root = LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            setBackgroundColor(NourishColors.SURFACE)
            setPadding(
                ui.dp(NourishSpacing.MD),
                ui.dp(NourishSpacing.SM),
                ui.dp(NourishSpacing.MD),
                0
            )
        }
        applySystemBarInsets(root)

        root.addView(headerPanel(nutritionMode))
        root.addView(tabRow())
        root.addView(headerStats(nutritionMode))

        val content = LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(0, ui.dp(NourishSpacing.SM), 0, ui.dp(NourishSpacing.XL))
        }
        val scrollView = ScrollView(activity).apply {
            isFillViewport = true
            addView(
                content,
                ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
            )
        }
        root.addView(
            scrollView,
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                0,
                1f
            )
        )

        activity.setContentView(root)
        root.requestApplyInsets()
        return content
    }

    private fun headerPanel(nutritionMode: Boolean): View = LinearLayout(activity).apply {
        orientation = LinearLayout.VERTICAL
        addView(brandBar())
        addView(profileRow(nutritionMode), matchWrapParams(topMargin = NourishSpacing.LG))
        addView(
            modeSwitchRow(),
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ui.dp(56)
            ).apply {
                topMargin = ui.dp(NourishSpacing.MD)
            }
        )
        layoutParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        ).apply {
            bottomMargin = ui.dp(NourishSpacing.SM)
        }
    }

    private fun brandBar(): LinearLayout = LinearLayout(activity).apply {
        orientation = LinearLayout.HORIZONTAL
        gravity = Gravity.CENTER_VERTICAL
        minimumHeight = ui.dp(48)

        val brand = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
        }
        val logo = ImageView(activity).apply {
            setImageResource(R.mipmap.ic_launcher)
            scaleType = ImageView.ScaleType.CENTER_CROP
            contentDescription = null
        }
        brand.addView(
            logo,
            LinearLayout.LayoutParams(ui.dp(28), ui.dp(28)).apply {
                rightMargin = ui.dp(NourishSpacing.XS)
            }
        )
        brand.addView(ui.displayText("NourishRx", NourishTypography.BODY_LARGE, NourishColors.GREEN_DARK))
        addView(brand, LinearLayout.LayoutParams(0, ui.dp(48), 1f))

        val showingAbout = callbacks.currentTab() == TAB_ABOUT
        val about = ui.button(
            "About",
            if (showingAbout) NourishColors.ON_ACCENT else NourishColors.INK_SECONDARY,
            if (showingAbout) NourishColors.BLUE else Color.TRANSPARENT
        ).apply {
            setOnClickListener { callbacks.selectAbout() }
        }
        addView(
            about,
            LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ui.dp(48)).apply {
                rightMargin = ui.dp(NourishSpacing.XS)
            }
        )

        val alerts = ui.button(callbacks.alertsLabel(), callbacks.alertColor(), Color.TRANSPARENT).apply {
            setOnClickListener { callbacks.handleAlertsTap() }
        }
        addView(alerts, LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ui.dp(48)))
    }

    private fun profileRow(nutritionMode: Boolean): LinearLayout = LinearLayout(activity).apply {
        orientation = LinearLayout.HORIZONTAL
        gravity = Gravity.CENTER_VERTICAL

        val profile = callbacks.selectedProfile()
        val identity = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            isClickable = true
            isFocusable = true
            minimumHeight = ui.dp(48)
            contentDescription = "Manage profile for ${profile.name}"
            setOnClickListener { callbacks.showProfiles() }
        }

        val avatarSize = 46
        val avatar = callbacks.profileAvatar(profile, avatarSize, NourishColors.GREEN, 16)
        identity.addView(
            avatar,
            LinearLayout.LayoutParams(
                ui.dp(callbacks.avatarWidthDp(profile, avatarSize)),
                ui.dp(avatarSize)
            ).apply {
                rightMargin = ui.dp(NourishSpacing.SM)
            }
        )

        val titleGroup = LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
        }
        val headline = ui.displayText(
            "${greeting()}, ${profile.name}",
            NourishTypography.TITLE,
            NourishColors.INK
        ).apply {
            setSingleLine(true)
            ellipsize = TextUtils.TruncateAt.END
        }
        titleGroup.addView(headline)
        titleGroup.addView(
            ui.text(
                LocalDate.now(zoneId).format(dateFormatter),
                NourishTypography.LABEL,
                NourishColors.MUTED,
                Typeface.NORMAL
            )
        )
        identity.addView(titleGroup, LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))

        val disclosure = ImageView(activity).apply {
            setImageResource(android.R.drawable.arrow_down_float)
            imageAlpha = 120
            contentDescription = null
        }
        identity.addView(disclosure, LinearLayout.LayoutParams(ui.dp(18), ui.dp(18)))
        addView(identity, LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))

        val quickAdd = ui.button(
            if (nutritionMode) "+ Log" else "+ Med",
            NourishColors.ON_ACCENT,
            NourishColors.GREEN
        ).apply {
            setOnClickListener { callbacks.showQuickAdd(nutritionMode) }
        }
        addView(
            quickAdd,
            LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ui.dp(48)).apply {
                leftMargin = ui.dp(NourishSpacing.SM)
            }
        )
    }

    private fun headerStats(nutritionMode: Boolean): LinearLayout {
        val stats = if (nutritionMode) nutritionStats() else medicationStats()
        return LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            setPadding(
                ui.dp(NourishSpacing.XS),
                ui.dp(NourishSpacing.SM),
                ui.dp(NourishSpacing.XS),
                ui.dp(NourishSpacing.SM)
            )
            background = ui.rounded(
                NourishColors.CARD,
                NourishColors.BORDER,
                ui.dp(NourishShapes.RADIUS_CARD)
            )
            stats.forEachIndexed { index, stat ->
                if (index > 0) {
                    addView(
                        View(activity).apply { setBackgroundColor(NourishColors.DIVIDER) },
                        LinearLayout.LayoutParams(ui.dp(1), ui.dp(44))
                    )
                }
                addView(statCell(stat), LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))
            }
            layoutParams = matchWrapParams(topMargin = NourishSpacing.SM)
        }
    }

    private fun medicationStats(): List<HeaderStat> {
        val medications = store.getAllMedications(callbacks.currentProfileId())
        val lowCount = medications.count { it.isLowStock() }
        val doses = callbacks.todayDoseCount()
        return listOf(
            HeaderStat(doses.toString(), if (doses == 1) "dose" else "doses", NourishColors.GREEN),
            HeaderStat(
                medications.size.toString(),
                if (medications.size == 1) "med" else "meds",
                NourishColors.BLUE
            ),
            HeaderStat(lowCount.toString(), if (lowCount == 1) "refill" else "refills", NourishColors.GOLD)
        )
    }

    private fun nutritionStats(): List<HeaderStat> {
        val today = LocalDate.now(zoneId)
        val start = today.atStartOfDay(zoneId).toInstant().toEpochMilli()
        val end = today.plusDays(1).atStartOfDay(zoneId).toInstant().toEpochMilli()
        val foodLogCount = store.getMealFoodLogs(callbacks.currentProfileId(), start, end).size
        val waterOunces = store.getWaterOunces(callbacks.currentProfileId(), start, end)
        val weights = store.getWeightEntries(callbacks.currentProfileId(), 1)
        return listOf(
            HeaderStat(
                foodLogCount.toString(),
                if (foodLogCount == 1) "food log" else "food logs",
                NourishColors.GREEN
            ),
            HeaderStat(waterOunces.toString(), "oz water", NourishColors.BLUE),
            HeaderStat(
                if (weights.isEmpty()) "--" else formatValue(weights[0].pounds),
                if (weights.isEmpty()) "weight" else "lb latest",
                NourishColors.GOLD
            )
        )
    }

    private fun statCell(stat: HeaderStat): LinearLayout = LinearLayout(activity).apply {
        orientation = LinearLayout.VERTICAL
        gravity = Gravity.CENTER
        addView(ui.displayText(stat.value, NourishTypography.TITLE, stat.color))
        addView(ui.text(stat.label, NourishTypography.CAPTION, NourishColors.MUTED, Typeface.NORMAL))
    }

    private fun modeSwitchRow(): LinearLayout = LinearLayout(activity).apply {
        orientation = LinearLayout.HORIZONTAL
        setPadding(
            ui.dp(NourishSpacing.XXS),
            ui.dp(NourishSpacing.XXS),
            ui.dp(NourishSpacing.XXS),
            ui.dp(NourishSpacing.XXS)
        )
        background = ui.rounded(
            NourishColors.SURFACE_SUBTLE,
            NourishColors.BORDER,
            ui.dp(NourishShapes.RADIUS_CONTROL)
        )
        addView(modeButton("Medication", MODE_MEDICATION))
        addView(modeButton("Nutrition", MODE_NUTRITION))
    }

    private fun modeButton(label: String, mode: String): Button {
        val selected = callbacks.currentMode() == mode
        return ui.button(
            label,
            if (selected) NourishColors.ON_ACCENT else NourishColors.INK_SECONDARY,
            if (selected) NourishColors.GREEN else Color.TRANSPARENT
        ).apply {
            setOnClickListener { callbacks.selectMode(mode) }
            layoutParams = LinearLayout.LayoutParams(0, ui.dp(48), 1f).apply {
                leftMargin = ui.dp(2)
                rightMargin = ui.dp(2)
            }
        }
    }

    private fun tabRow(): LinearLayout {
        val tabs = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
        }
        if (callbacks.currentMode() == MODE_NUTRITION) {
            tabs.addView(tabItem("Today", "nutrition_today"))
            tabs.addView(tabItem("Meals", "nutrition_meals"))
            tabs.addView(tabItem("Saved", "nutrition_saved"))
            tabs.addView(tabItem("Foods", "nutrition_foods"))
            tabs.addView(tabItem("Body", "nutrition_body"))
        } else {
            tabs.addView(tabItem("Today", "today"))
            tabs.addView(tabItem("Medications", "meds"))
            tabs.addView(tabItem("Inventory", "stock"))
        }

        return LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            addView(
                tabs,
                LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ui.dp(48))
            )
            addView(
                View(activity).apply { setBackgroundColor(NourishColors.DIVIDER) },
                LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ui.dp(1))
            )
        }
    }

    private fun tabItem(label: String, tab: String): LinearLayout {
        val selected = callbacks.currentTab() == tab
        val labelView = ui.text(
            label,
            NourishTypography.CAPTION,
            if (selected) NourishColors.GREEN_DARK else NourishColors.INK_SECONDARY,
            if (selected) Typeface.BOLD else Typeface.NORMAL
        ).apply {
            gravity = Gravity.CENTER
            setSingleLine(true)
            ellipsize = TextUtils.TruncateAt.END
            setPadding(ui.dp(NourishSpacing.XXS), 0, ui.dp(NourishSpacing.XXS), 0)
        }
        val indicator = View(activity).apply {
            setBackgroundColor(if (selected) NourishColors.GREEN else Color.TRANSPARENT)
        }

        return LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER
            isClickable = true
            isFocusable = true
            isSelected = selected
            contentDescription = if (selected) "$label tab, selected" else "$label tab"
            val selectableBackground = TypedValue()
            if (activity.theme.resolveAttribute(
                    android.R.attr.selectableItemBackground,
                    selectableBackground,
                    true
                )
            ) {
                setBackgroundResource(selectableBackground.resourceId)
            }
            setOnClickListener { callbacks.selectTab(tab) }
            addView(labelView, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f))
            addView(
                indicator,
                LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ui.dp(3)).apply {
                    leftMargin = ui.dp(NourishSpacing.SM)
                    rightMargin = ui.dp(NourishSpacing.SM)
                }
            )
            layoutParams = LinearLayout.LayoutParams(0, ui.dp(48), 1f)
        }
    }

    @Suppress("DEPRECATION")
    private fun applySystemBarInsets(root: View) {
        root.setOnApplyWindowInsetsListener { view, insets ->
            val topInset: Int
            val bottomInset: Int
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val systemBars = insets.getInsets(WindowInsets.Type.systemBars())
                topInset = systemBars.top
                bottomInset = systemBars.bottom
            } else {
                topInset = insets.systemWindowInsetTop
                bottomInset = insets.systemWindowInsetBottom
            }
            view.setPadding(
                ui.dp(NourishSpacing.MD),
                topInset + ui.dp(NourishSpacing.SM),
                ui.dp(NourishSpacing.MD),
                bottomInset
            )
            insets
        }
    }

    private fun greeting(): String = when (LocalTime.now(zoneId).hour) {
        in 5..11 -> "Good morning"
        in 12..16 -> "Good afternoon"
        else -> "Good evening"
    }

    private fun matchWrapParams(topMargin: Int = 0): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        ).apply {
            this.topMargin = ui.dp(topMargin)
        }

    private fun formatValue(value: Float): String {
        val rounded = value.roundToInt()
        return if (abs(value - rounded) < 0.05f) {
            rounded.toString()
        } else {
            String.format(Locale.getDefault(), "%.1f", value)
        }
    }

    private companion object {
        const val MODE_MEDICATION = "medication"
        const val MODE_NUTRITION = "nutrition"
        const val TAB_ABOUT = "about"
    }
}
